pkg/cmdutil/errors.go

```diff
@@ -57,6 +58,21 @@ func MutuallyExclusive(message string, conditions ...bool) error {
 	return nil
 }
 
+func ValidateHasOnlyOneRemote(hasRepoOverride bool, remotes func() (ghContext.Remotes, error)) error {
+	if !hasRepoOverride && remotes != nil {
+		remotes, err := remotes()
+		if err != nil {
+			return err
+		}
+
+		if remotes.Len() > 1 {
+			return fmt.Errorf("multiple remotes detected %v. please specify which repo to use by providing the -R or --repo argument", remotes)
+		}
+	}
+
+	return nil
+}
```

Comment:
Gentle nudge on this @wingleung, don't want it to get stale.

@andyfeller if I read this correctly it's a breaking change for the following **non-interactive** script:

```
gh repo create org/repo
gh repo fork org/repo
gh repo clone me/fork <-- sets the default to org/repo
gh secret set FOO BAR
```

Previously this would have put the secret into upstream (the bug), but now it will error. Just want to make sure we are clear and can communicate this.

Reply:
I can't see any way to avoid it being a breaking change while also helping our users avoid a footgun. I'm a bit concerned about how many people it might break though.

https://github.com/search?type=code&q=%22gh+secret+set%22

More than 1.7 public results, and any of these that are run non-interactively against forks will be broken.

Reply:
I've been very busy at work so apologies for the late reply.

OK, so if I understand correctly, there are 3 flows we want to cover, 2 of which I already implemented in this PR

### ✅ flow 1: command with 1 remote
just use that single remote and proceed with the code

### ✅ flow 2: command with multiple remotes in non-interactive mode
error out but provide documentation on how to force a remote with `--repo`.

> This would mean we don't need users to provide -R,--repo override as the previous workaround involved.

in this case we would still need to support `-R` and `--repo` overrides, otherwise non-interactive flow will get stuck, or am I not interpreting it correctly? 😅 

```bash
multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
```

**use cases**

- `gh secret set MYSECRET --body "$ENV_VALUE”`
- `gh variable set MYVARIABLE --body "$ENV_VALUE”`

### 🚫 flow 3: command with multiple remotes in interactive mode
show a prompt to select 1 remote repo from multiple available remote repos using a search prompt 

**use cases**

- `gh secret list`
- `gh secret set MYSECRET`
- `gh variable set MYVARIABLE`

Is that correct?

Reply:
### Breaking changes

we could do a release we introduce a warning for the v2 cli and mention this would change in v3 into a breaking change?

Reply:
nice to see how remotes are being handle in other places, maybe in another ticket it could be looked into on how to merge all those strategies into 1 helper function.

- default to `origin`, fallback to first remote 👉 https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/extension/manager.go#L691-L716

- default remote set to configured `set-default` remote value with a fallback to the first remote and prompt for user input on which remote to pick 👉 https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/codespace/create.go#L143-L174

just to prevent this PR from getting too big or out of scope, I'd suggest to do this in another ticket 😄

Reply:
> I've been very busy at work so apologies for the late reply.

@wingleung : I just want to say you owe us no apology; thank you for yet again working to improve `gh` ❤️

Reply:
I think this summarizes the changes adequately, yes:

> ✅ flow 1: command with 1 remote
> ✅ flow 2: command with multiple remotes in non-interactive mode
> 🚫 flow 3: command with multiple remotes in interactive mode

Along with the actual changes to the command, we will want to make sure to call this out in the release notes to explain the motivation and hopefully how this is a one-time issue for non-interactive situations.